jgrapht-core/src/main/java/org/jgrapht/alg/lca/EulerTourRMQLCAFinder.java

```diff
@@ -0,0 +1,283 @@
+/*
+ * (C) Copyright 2018-2018, by Alexandru Valeanu and Contributors.
+ *
+ * JGraphT : a free Java graph-theory library
+ *
+ * This program and the accompanying materials are dual-licensed under
+ * either
+ *
+ * (a) the terms of the GNU Lesser General Public License version 2.1
+ * as published by the Free Software Foundation, or (at your option) any
+ * later version.
+ *
+ * or (per the licensee's choosing)
+ *
+ * (b) the terms of the Eclipse Public License v1.0 as published by
+ * the Eclipse Foundation.
+ */
+package org.jgrapht.alg.lca;
+
+import org.jgrapht.Graph;
+import org.jgrapht.Graphs;
+import org.jgrapht.alg.interfaces.LowestCommonAncestorAlgorithm;
+import org.jgrapht.alg.util.Pair;
+import org.jgrapht.util.VertexToIntegerMapping;
+
+import java.util.*;
+
+import static org.jgrapht.util.MathUtil.log2;
+
+/**
+ * Algorithm for computing lowest common ancestors in rooted trees and forests based on
+ * <i>Berkman, Omer; Vishkin, Uzi (1993), "Recursive Star-Tree Parallel Data Structure",
+ * SIAM Journal on Computing, 22 (2): 221–242, doi:10.1137/0222017</i>.
+ *
+ * <p>
+ *  The algorithm involves forming an Euler tour of a graph formed from the input tree by doubling every edge,
+ *  and using this tour to compute a sequence of level numbers of the nodes in the order the tour visits them.
+ *  A lowest common ancestor query can then be transformed into a query that seeks the minimum value occurring
+ *  within some subinterval of this sequence of numbers.
+ * </p>
+ *
+ * <p>
+ *  Preprocessing Time complexity: $O(|V| log(|V|))$<br>
+ *  Preprocessing Space complexity:  $O(|V| log(|V|))$<br>
+ *  Query Time complexity: $O(1)$<br>
+ *  Query Space complexity: $O(1)$<br>
+ * </p>
+ *
+ * @param <V> the graph vertex type
+ * @param <E> the graph edge type
+ *
+ * @author Alexandru Valeanu
+ */
+public class EulerTourRMQLCAFinder<V, E> implements LowestCommonAncestorAlgorithm<V> {
+    private final Graph<V, E> graph;
+    private final Set<V> roots;
+    private final int maxLevel;
+
+    private Map<V, Integer> vertexMap;
+    private List<V> indexList;
+
+    private int[] eulerTour;
+    private int sizeTour;
+
+    private int numberComponent;
+    private int[] component;
+
+    private int[] level;
+    private int[] representative;
+
+    private int[][] rmq;
+    private int[] log2;
```

Comment:
It's kind of confusing to name this array log2 when you've already done a static import of MathUtil.log2.